modules/diskq/diskq-serializer.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2020 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "diskq-serializer.h"
+
+#include <unistd.h>
+
+typedef struct _MsgSerializer
+{
+  SerializeArchive super;
+  int fd;
+  gsize offset;
+} MsgSerializer;
+
+static ssize_t
+_pread_all(int fd, char *buf,  size_t count, off_t offset)
+{
+  int retval = pread(fd, buf, count, offset);
+
+  if (retval <= 0)
+    return retval;
+
+  if (retval == count)
+    return count;
+
+  return retval + _pread_all(fd, buf+retval, count-retval, offset+retval);
```

Comment:
What if the `retval` is `10` and the next `pread` call returns `-1`? We return `9` in that case, which is not good.

Reply:
I think a simple while loop would be better in this case.

Reply:
Fixed thanks.